server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java

```diff
@@ -0,0 +1,321 @@
+import org.opensearch.common.settings.ClusterSettings;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.transport.TransportException;
+import org.opensearch.transport.TransportResponse;
+import org.opensearch.transport.TransportResponseHandler;
+import org.opensearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Predicate;
+
+public class DecommissionService implements ClusterStateApplier {
+
+    private static final Logger logger = LogManager.getLogger(DecommissionService.class);
+
+    private final ClusterService clusterService;
+    private final TransportService transportService;
+    private final NodeRemovalClusterStateTaskExecutor nodeRemovalExecutor;
+    private final ThreadPool threadPool;
+    private final DecommissionHelper decommissionHelper;
+    private ClusterState clusterState;
+    private volatile List<String> awarenessAttributes;
+
+    @Inject
+    public DecommissionService(
+        Settings settings,
+        ClusterSettings clusterSettings,
+        ClusterService clusterService,
+        TransportService transportService,
+        ThreadPool threadPool,
+        AllocationService allocationService
+    ) {
+        this.clusterService = clusterService;
+        this.transportService = transportService;
+        this.threadPool = threadPool;
+        this.nodeRemovalExecutor = new NodeRemovalClusterStateTaskExecutor(allocationService, logger);
+        this.decommissionHelper = new DecommissionHelper(
+            clusterService,
+            nodeRemovalExecutor
+        );
+        this.awarenessAttributes = AwarenessAllocationDecider.CLUSTER_ROUTING_ALLOCATION_AWARENESS_ATTRIBUTE_SETTING.get(settings);
+        clusterSettings.addSettingsUpdateConsumer(
+            AwarenessAllocationDecider.CLUSTER_ROUTING_ALLOCATION_AWARENESS_ATTRIBUTE_SETTING,
+            this::setAwarenessAttributes
+        );
+    }
+
+    List<String> getAwarenessAttributes() {
+        return awarenessAttributes;
+    }
+
+    private void setAwarenessAttributes(List<String> awarenessAttributes) {
+        this.awarenessAttributes = awarenessAttributes;
+    }
+
+    public void initiateAttributeDecommissioning(
+        final DecommissionAttribute decommissionAttribute,
+        final ActionListener<ClusterStateUpdateResponse> listener,
+        ClusterState state
+    ) {
+        /*
+         * 1. Abdicate master
+         * 2. Register attribute -> status should be set to INIT
+         * 3. Trigger weigh away for graceful decommission -> status should be set to DECOMMISSIONING
+         * 4. Once zone is weighed away -> trigger zone decommission using executor -> status should be set to DECOMMISSIONED on successful response
+         * 5. Clear voting config
+         */
+        this.clusterState = state;
```

Comment:
A method should not be updating the state (here - clusterState) of a service. If the service's scope is only per request, can we make sure that the clusterState is set in the constructor and set as final. If there is a scope of concurrency here, then pls handle using some synchronisation and discarding older cluster states.

Reply:
Ack. Will remove this in next commit.